chia/full_node/full_node.py

```diff
@@ -709,101 +709,100 @@ async def _sync(self):
+                        if peer.closed:
+                            peers_with_peak.remove(peer)
+                            continue
+                        response = await peer.request_blocks(request, timeout=10)
+                        if response is None:
+                            await peer.close()
+                            peers_with_peak.remove(peer)
+                        elif isinstance(response, RespondBlocks):
+                            await batch_queue.put((peer, response.blocks))
+                            fetched = True
+                            break
+                    if fetched is False:
+                        self.log.error(f"failed fetching {start_height} to {end_height} from peers")
+                        await batch_queue.put(None)
+                        return
+                    if self.sync_store.peers_changed.is_set():
+                        peers_with_peak = self.get_peers_with_peak(peak_hash)
+                        self.sync_store.peers_changed.clear()
+            except Exception as e:
+                self.log.error(f"Exception fetching {start_height} to {end_height} from peer {e}")
+            finally:
+                # finished signal with None
+                await batch_queue.put(None)
+
+        async def validate_block_batches(batch_queue):
+            advanced_peak = False
+            while True:
+                res = await batch_queue.get()
+                if res is None:
+                    self.log.debug("done fetching blocks")
+                    return
+                peer, blocks = res
+                start_height = blocks[0].height
+                end_height = blocks[-1].height
+                success, advanced_peak, _ = await self.receive_block_batch(
+                    blocks, peer, None if advanced_peak else uint32(fork_point_height), summaries
+                )
+                if success is False:
+                    if peer in peers_with_peak:
+                        peers_with_peak.remove(peer)
+                    await peer.close(600)
+                    raise ValueError(f"Failed to validate block batch {start_height} to {end_height}")
+                self.log.info(f"Added blocks {start_height} to {end_height}")
+                await self.send_peak_to_wallets()
+                self.blockchain.clean_block_record(end_height - self.constants.BLOCKS_CACHE_SIZE)
 
-            for peer in to_remove:
-                peers_with_peak.remove(peer)
+        loop = asyncio.get_event_loop()
+        batch_queue: asyncio.Queue = asyncio.Queue(loop=loop, maxsize=buffer_size)
+        fetch_task = asyncio.Task(fetch_block_batches(batch_queue, peers_with_peak))
+        validate_task = asyncio.Task(validate_block_batches(batch_queue))
+        try:
+            await asyncio.gather(fetch_task, validate_task)
+        except Exception as e:
+            fetch_task.cancel()
```

Comment:
it seems like `validate_task` should be cancelled here too, no?
and probably both `fetch_task` and `validate_task` should be joined still. It seems like a serious error to fail to join them though. Is there a known reason for the `gather()` to fail?

I imagine the `raise ValueError()` would exit here. Presumably there's a risk that the `fetch_task` throws as well.

Reply:
i don't think we need to cancel validate_task, fetch_task finally clause submits  None to the queue which will cause validate_task to shut down gracefully, gather would throw an error if we had an exception in one of validate_task/fetch_task. i don't mind adding the cancel but i was thinking its better let validate exit in a clean way.

Reply:
right, if `fetch_task` throws, it will still push `None` onto the queue, to signal `validate_task` to exit. I would think we should still "join" the task though. Otherwise we might end up launching another `validate_task` before the previous one exits

Reply:
I think this is still an issue. Apparently you can configure `gather()` to not abort on exceptions, in which case you will correctly join the `validate_task` as well. Then you wouldn't get the exception though. I would think you should just add:
```
await fetch_task
await validate_task
```

In the exception case. (at least I would expect you can wait on a handle to a task that has already completed

Reply:
Hi almog I think this is the last issue for it to go in

Reply:
maybe im missing something but i don't really understand why this is needed, fetch_task cant should never throw we added a try catch there just for that, so this should always exit gracefully

Reply:
I see. The risk I see is that the correctness of the `gather()` call site depends on implementation details of the fetch task and the validate task. In the future, changes might be made to this code, that appear correct, that violate this expectation and introduce a bug.

When just reading the code around `gather()` it looks like tasks can leak. When just reading the fetch-task code, it's not obvious that preventing such leak depends on it catching all exceptions (and that pushing `None` onto the queue also may not raise an exception).

I can't think of a simple change to improve this situation, but I think it should at least be explained in a comment and ideally with some `assert`s as well. For example, you expect that the `validate_task` is always done by the time `gather()` returns (or an exception comes out of it). That could be asserted.

Reply:
added a comment and assert